ambry-cloud/src/main/java/com/github/ambry/cloud/CloudStorageCompactor.java

```diff
@@ -103,47 +106,80 @@ public int compactPartitions() {
       logger.info("Skipping compaction as no partitions are assigned.");
       return 0;
     }
+    List<PartitionId> partitionSnapshot = new ArrayList<>(partitions);
+    long compactionStartTime = System.currentTimeMillis();
+    long timeToQuit = System.currentTimeMillis() + compactionTimeLimitMs;
+    int compactionInProgress = 0;
+    doneLatch.set(new CountDownLatch(1));
+    int totalBlobsPurged = 0;
+    int compactedPartitionCount = 0;
+    try {
+      while (true) {
+        while (compactionInProgress < numThreads) {
+          if (partitionSnapshot.isEmpty()) {
+            break;
+          }
+          PartitionId partitionId = partitionSnapshot.remove(0);
+          executorCompletionService.submit(() -> compactPartition(partitionId));
+          compactionInProgress++;
+        }
+        totalBlobsPurged += executorCompletionService.take().get();
+        compactionInProgress--;
+        compactedPartitionCount++;
+        if (System.currentTimeMillis() >= timeToQuit) {
+          logger.info("Compaction terminated due to time limit exceeded.");
+          break;
+        }
+        if (isShuttingDown()) {
+          logger.info("Compaction terminated due to shut down.");
+          break;
+        }
+        if (partitionSnapshot.isEmpty()) {
+          break;
+        }
+      }
+      while (compactionInProgress > 0) {
+        totalBlobsPurged += executorCompletionService.take().get();
```

Comment:
Is the idea to wait until all the threads are done before assigning any new work?  That won't be optimal since a long running partition can keep the other threads idle for a while.

Reply:
No.. the idea is not to wait until all threads are done. The idea to wait for any one of the threads to return, and then assign it new work immediately. `executorCompletionService.take()` retrieves and removes the `Future` representing the next completed task, waiting if none are yet present. Calling `.get()` on the future just returns the result.

So basically first we assign work to all threads. Then wait for any single thread to complete its work in the `executorCompletionService.take().get()` call. And then we loop over to assign work to this thread.

Reply:
Note that the final while loop `while (compactionInProgress > 0)` does loop through all the threads. This happens when all the partitions have been assigned, and we have come out of the initial `while(true)` loop. Here we are just making sure that all the threads get completed before ending the compaction run.

Reply:
Note that the final while loop `while (compactionInProgress > 0)` does loop through all the threads. This happens when all the partitions have been assigned, and we have come out of the initial `while(true)` loop. Here we are just making sure that all the threads get completed before ending the compaction run.

This loop also takes care of the case where we encounter a compaction shutdown. In that case too, we break out of the `while(true)` loop, and wait for all the threads to complete work, before releasing the done latch....

Reply:
Okay, I see that now.